url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/PartsRobotAgent.java

```diff
@@ -0,0 +1,531 @@
+
+	private List<Kit> KitsOnStand;
+
+	private Stand stand;
+	private PartsRobotGraphics partsRobotGraphics;
+
+	private float dropChance = 0;
+
+	public class MyKit {
+		public Kit kit;
+		public MyKitStatus MKS;
+
+		public MyKit(Kit k) {
+			kit = k;
+			MKS = MyKitStatus.NOT_DONE;
+		}
+	}
+
+	public enum MyKitStatus {
+		NOT_DONE, DONE
+	};
+
+	public class Arm {
+		Part part;
+		ArmStatus AS;
+
+		public Arm() {
+			part = null;
+			AS = ArmStatus.EMPTY;
+		}
+	}
+
+	private enum ArmStatus {
+		EMPTY, FULL, EMPTYING
+	};
+
+	private PartsRobotStatus status;
+
+	private enum PartsRobotStatus {
+		IDLE, PICKING_UP, PLACING
+	};
+
+	public Semaphore animation = new Semaphore(0, true);
+
+	// public Semaphore accessKit = new Semaphore(0, true);
+
+	public PartsRobotAgent(String name) {
+		super();
+
+		this.name = name;
+		time = new Date();
+		timer = new Timer();
+
+		// Add arms
+		for (int i = 0; i < 4; i++) {
+			this.Arms.add(new Arm());
+		}
+
+		status = PartsRobotStatus.IDLE;
+	}
+
+	/*
+	 * Messages
+	 */
+
+	@Override
+	public void msgSetDropChance(float dChance) {
+		print("Chance to drop is now set to: " + dChance);
+		dropChance = dChance;
+	}
+
+	/**
+	 * Changes the configuration for the kits From FCS
+	 */
+	@Override
+	public void msgHereIsKitConfiguration(KitConfig config) {
+		print("Received msgHereIsKitConfiguration");
+		Kitconfig = config;
+		kitsNum = 0;
+		GoodParts = new ConcurrentHashMap<Nest, List<Part>>();
+		// stateChanged();
+	}
+
+	/**
+	 * From Camera
+	 */
+	@Override
+	public void msgHereAreGoodParts(Nest n, List<Part> goodParts) {
+		print("Received msgHereAreGoodParts of type " + goodParts.get(0).type.getName());
+		GoodParts.put(n, goodParts);
+		print("I have " + MyKits.size() + " kits and " + GoodParts.size() + " nests");
+		stateChanged();
+	}
+
+	/**
+	 * From Stand
+	 */
+	@Override
+	public void msgUseThisKit(final Kit k) {
+		print("Received msgUseThisKit");
+
+		MyKit mk = new MyKit(k);
+		MyKits.add(mk);
+		print("I have " + MyKits.size() + " kits and " + GoodParts.size() + " nests");
+		stateChanged();
+
+	}
+
+	/**
+	 * Releases animation semaphore after a part is picked up, so that a new animation may be run by graphics. From
+	 * graphics
+	 */
+	@Override
+	public void msgPickUpPartDone() {
+		print("Received msgPickUpPartDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	/**
+	 * Releases animation semaphore after a part is given to kit, so that a new animation may be run by graphics. From
+	 * graphics
+	 */
+	@Override
+	public void msgGivePartToKitDone() {
+		print("Received msgGivePartToKitDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgDropPartFromArmDone() {
+		print("Received msgDropPartFromArmDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	/*
+	 * Scheduler
+	 * 
+	 * @see agent.Agent#pickAndExecuteAnAction()
+	 */
+
+	@Override
+	public boolean pickAndExecuteAnAction() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 188-257 into a separate method.